diff --git a/CRM/Contribute/Form/ContributionPage/Settings.php b/CRM/Contribute/Form/ContributionPage/Settings.php
index b6991b9fdf..aed5c79ac2 100644
--- a/CRM/Contribute/Form/ContributionPage/Settings.php
+++ b/CRM/Contribute/Form/ContributionPage/Settings.php
@@ -168,7 +168,7 @@ class CRM_Contribute_Form_ContributionPage_Settings extends CRM_Contribute_Form_
     $this->addElement('checkbox', 'is_confirm_enabled', ts('Use a confirmation page?'));
 
     // is this page shareable through social media ?
-    $this->addElement('checkbox', 'is_share', ts('Add footer region with Twitter, Facebook and LinkedIn share buttons and scripts?'));
+    $this->addElement('checkbox', 'is_share', ts('Add footer region with Twitter, Facebook and Instagram share buttons and scripts?'));
 
     // is this page active ?
     $this->addElement('checkbox', 'is_active', ts('Is this Online Contribution Page Active?'));
diff --git a/templates/CRM/common/SocialNetwork.tpl b/templates/CRM/common/SocialNetwork.tpl
index c1b13cac34..5ba9622b72 100644
--- a/templates/CRM/common/SocialNetwork.tpl
+++ b/templates/CRM/common/SocialNetwork.tpl
@@ -22,11 +22,11 @@
     {if $emailMode eq true}
         <a href="https://twitter.com/share?url={$url|escape:'url'}&amp;text={$title|escape:'url'}" class="btn btn-default" role="button" target="_blank" title="{ts escape='htmlattribute'}Tweet{/ts}">{ts}Twitter{/ts}</a>
         <a href="https://facebook.com/sharer/sharer.php?u={$url|escape:'url'}" target="_blank" class="btn btn-default" role="button" title="{ts escape='htmlattribute'}Share{/ts}">{ts}Facebook{/ts}</a>
+        <a href="https://www.instagram.com/?url={$url|escape:'url'}&amp;title={$title|escape:'url'}" target="_blank" rel="noopener" class="btn btn-default" title="{ts escape='htmlattribute'}Share{/ts}">{ts}Instagram{/ts}</a>
-        <a href="https://www.linkedin.com/shareArticle?mini=true&amp;url={$url|escape:'url'}&amp;title={$title|escape:'url'}" target="_blank" rel="noopener" class="btn btn-default" title="{ts escape='htmlattribute'}Share{/ts}">{ts}LinkedIn{/ts}</a>
     {else}
         <button onclick="window.open('https://twitter.com/intent/tweet?url={$url|escape:'url'}&amp;text={$title|escape:'url'}','_blank')" type="button" class="btn btn-default crm-button" id="crm-tw" title="{ts escape='htmlattribute'}Tweet{/ts}"><i aria-hidden="true" class="crm-i fa-twitter"></i>&nbsp;&nbsp;{ts}Twitter{/ts}</button>
         <button onclick="window.open('https://facebook.com/sharer/sharer.php?u={$url|escape:'url'}','_blank')" type="button" class="btn btn-default crm-button" role="button" id="crm-fb" title="{ts escape='htmlattribute'}Share{/ts}"><i aria-hidden="true" class="crm-i fa-facebook"></i>&nbsp;&nbsp;{ts}Facebook{/ts}</button>
+        <button onclick="window.open('https://www.instagram.com/?url={$url|escape:'url'}&amp;title={$title|escape:'url'}','_blank')" type="button" rel="noopener" class="btn btn-default crm-button" id="crm-li" title="{ts escape='htmlattribute'}Share{/ts}"><i aria-hidden="true" class="crm-i fa-instagram"></i>&nbsp;&nbsp;{ts}Instagram{/ts}</button>
-        <button onclick="window.open('https://www.linkedin.com/shareArticle?mini=true&amp;url={$url|escape:'url'}&amp;title={$title|escape:'url'}','_blank')" type="button" rel="noopener" class="btn btn-default crm-button" id="crm-li" title="{ts escape='htmlattribute'}Share{/ts}"><i aria-hidden="true" class="crm-i fa-linkedin"></i>&nbsp;&nbsp;{ts}LinkedIn{/ts}</button>
         <button onclick="window.open('mailto:?subject={$title|escape:'quotes'}&amp;body={$url|escape:'url'}','_self')" type="button" rel="noopener" class="btn btn-default crm-button" id="crm-email"><i aria-hidden="true" class="crm-i fa-envelope" title="{ts escape='htmlattribute'}Email{/ts}"></i>&nbsp;&nbsp;{ts}Email{/ts}</button>
     {/if}
     {if $pageURL}

