From f7957bebd3cefc89a218e16dca983e6ab0a8a7fc Mon Sep 17 00:00:00 2001
From: gitressa <3491208+gitressa@users.noreply.github.com>
Date: Thu, 9 Jan 2025 18:19:49 -0500
Subject: [PATCH] Missing route cachability metadata in CiviCRM

---
 src/CivicrmBreadcrumbBuilder.php | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/CivicrmBreadcrumbBuilder.php b/src/CivicrmBreadcrumbBuilder.php
index f36d97d..a91d675 100644
--- a/src/CivicrmBreadcrumbBuilder.php
+++ b/src/CivicrmBreadcrumbBuilder.php
@@ -4,6 +4,7 @@
 
 use Drupal\Core\Breadcrumb\BreadcrumbBuilderInterface;
 use Drupal\Core\Breadcrumb\Breadcrumb;
+use Drupal\Core\Cache\CacheableMetadata;
 use Drupal\Core\Routing\RouteMatchInterface;
 use Drupal\Core\Link;
 use Drupal\Core\StringTranslation\StringTranslationTrait;
@@ -30,9 +31,13 @@ public function __construct(TranslationInterface $stringTranslation, CivicrmPage
   /**
    * {@inheritdoc}
    */
-  public function applies(RouteMatchInterface $route_match) {
+  public function applies(RouteMatchInterface $route_match, ?CacheableMetadata $cacheable_metadata = NULL) {
     $route_object = $route_match->getRouteObject();
 
+    if ($cacheable_metadata) {
+      $cacheable_metadata->addCacheContexts(['route']);
+    }
+
     // No route object is defined, so we can't inspect it.
     if (!$route_object) {
       return FALSE;
