diff --git a/js/fullcalendar_view.js b/js/fullcalendar_view.js
index 80cd14438..071d30cef 100644
--- a/js/fullcalendar_view.js
+++ b/js/fullcalendar_view.js
@@ -164,6 +164,9 @@
       //Create window
       dialogs[dialogIndex] = jsFrame.create(dialogOptions);
       
+      let jsFrameEvent = $.Event('jsframe_create');
+      jsFrameEvent.jsFrame = jsFrame;
+      $.trigger(jsFrameEvent);
       dialogs[dialogIndex].show();
       dialogIndex++;
 
