diff --git a/web/sites/default/modules/webform_civicrm/src/WebformCivicrmPostProcess.php b/web/sites/default/modules/webform_civicrm/src/WebformCivicrmPostProcess.php
index 00707d895..81faab1a1 100755
--- a/web/sites/default/modules/webform_civicrm/src/WebformCivicrmPostProcess.php
+++ b/web/sites/default/modules/webform_civicrm/src/WebformCivicrmPostProcess.php
@@ -167,6 +167,9 @@ protected function modifyWebformSubmissionData(WebformSubmissionInterface $webfo
       if ($element && $element['#type'] == 'civicrm_options' && is_array($val) && count(array_filter(array_keys($val), 'is_string')) > 0) {
         $data[$field_key] = array_values($val);
       }
+      if (empty($data[$field_key]) && !empty($_POST[$field_key])) {
+         $data[$field_key] = $_POST[$field_key];
+      }
     }
     $webform_submission->setData($data);
   }
